File: bindings/python/tests/test_basic.py
"""Basic encode/decode + lifecycle tests."""

from __future__ import annotations

import pytest

import ztok


def test_version_is_nonempty_string() -> None:
    v = ztok.version()
    assert isinstance(v, str)
    assert v.count(".") >= 1
    assert all(part.isdigit() for part in v.split(".")[:2])


def test_byte_id_encode_decode_roundtrip() -> None:
    with ztok.Pipeline.byte_id() as pipe:
        ids = pipe.encode("hi")
        assert ids == [0x68, 0x69]
        assert pipe.decode(ids) == "hi"


def test_bpe_encode_decode_roundtrip(bpe_pipeline: ztok.Pipeline) -> None:
    text = "hello world"
    ids = bpe_pipeline.encode(text)
    assert len(ids) > 0
    assert bpe_pipeline.decode(ids) == text


def test_decoded_text_equals_input_on_100_lines(bpe_pipeline: ztok.Pipeline) -> None:
    # 100 lines: alternating short/medium English snippets from the vocab's
    # merge coverage so we don't blow byte-fallback budget on novel UTF-8.
    snippets = [
        "hello world",
        "the quick brown fox",
        "foo bar baz",
        "hello there hello world",
        "the the the",
        "  the  ",
        "foo",
        "hello",
        " world",
        "bar baz",
    ]
    lines = [snippets[i % len(snippets)] for i in range(100)]
    for line in lines:
        ids = bpe_pipeline.encode(line)
        decoded = bpe_pipeline.decode(ids)
        assert decoded == line, f"round-trip failed for {line!r}: got {decoded!r}"


def test_empty_input_returns_empty_ids() -> None:
    with ztok.Pipeline.byte_id() as pipe:
        assert pipe.encode("") == []
        assert pipe.decode([]) == ""


def test_context_manager_closes_pipeline() -> None:
    with ztok.Pipeline.byte_id() as pipe:
        assert pipe.encode("x") == [ord("x")]
    # After exit the handle must be released. Calling encode raises.
    with pytest.raises(ztok.ZtokError):
        pipe.encode("y")


def test_explicit_close_is_idempotent() -> None:
    pipe = ztok.Pipeline.byte_id()
    pipe.close()
    pipe.close()  # second close must be a no-op
    with pytest.raises(ztok.ZtokError):
        pipe.encode("z")


def test_invalid_input_raises_typed_error() -> None:
    # Passing an unknown model kind through the raw config trips
    # ZTOK_ERR_INVALID_INPUT -> ZtokInvalidInputError.
    with pytest.raises(ztok.ZtokInvalidInputError):
        ztok.Pipeline.byte_id(normalizer=999)  # type: ignore[arg-type]


def test_decode_bytes_returns_raw_bytes() -> None:
    with ztok.Pipeline.byte_id() as pipe:
        ids = pipe.encode("ab")
        assert pipe.decode_bytes(ids) == b"ab"

File: bindings/ruby/test/test_basic.rb
# frozen_string_literal: true

require_relative "helper"

class TestBasic < Minitest::Test
  include ZtokTestFixtures

  def test_version_is_nonempty_string
    v = Ztok.version
    assert_kind_of String, v
    assert_includes v, "."
    major, minor = v.split(".")[0, 2]
    assert_match(/^\d+$/, major)
    assert_match(/^\d+$/, minor)
  end

  def test_byte_id_encode_decode_roundtrip
    pipe = Ztok::Pipeline.byte_id
    begin
      ids = pipe.encode("hi")
      assert_equal [0x68, 0x69], ids
      assert_equal "hi", pipe.decode(ids)
    ensure
      pipe.close
    end
  end

  def test_bpe_encode_decode_roundtrip
    pipe = Ztok::Pipeline.from_tiktoken(ZtokTestFixtures.tiktoken_fixture, cl100k: true)
    begin
      text = "hello world"
      ids = pipe.encode(text)
      assert_operator ids.length, :>, 0
      assert_equal text, pipe.decode(ids)
    ensure
      pipe.close
    end
  end

  def test_100_line_roundtrip_stress
    pipe = Ztok::Pipeline.from_tiktoken(ZtokTestFixtures.tiktoken_fixture, cl100k: true)
    snippets = [
      "hello world",
      "the quick brown fox",
      "foo bar baz",
      "hello there hello world",
      "the the the",
      "  the  ",
      "foo",
      "hello",
      " world",
      "bar baz",
    ]
    begin
      100.times do |i|
        line = snippets[i % snippets.length]
        ids = pipe.encode(line)
        assert_equal line, pipe.decode(ids), "round-trip failed for #{line.inspect}"
      end
    ensure
      pipe.close
    end
  end

  def test_empty_input_returns_empty_ids
    pipe = Ztok::Pipeline.byte_id
    begin
      assert_equal [], pipe.encode("")
      assert_equal "", pipe.decode([])
    ensure
      pipe.close
    end
  end

  def test_close_is_idempotent_and_postclose_ops_raise
    pipe = Ztok::Pipeline.byte_id
    pipe.close
    pipe.close # second close must be a no-op
    assert pipe.closed?
    err = assert_raises(Ztok::Error) { pipe.encode("z") }
    assert_match(/closed/i, err.message)
  end

  def test_invalid_input_raises_typed_error
    assert_raises(Ztok::InvalidInputError) do
      Ztok::Pipeline.byte_id(normalizer: 999)
    end
  end

  def test_decode_bytes_returns_raw_bytes
    pipe = Ztok::Pipeline.byte_id
    begin
      ids = pipe.encode("ab")
      raw = pipe.decode_bytes(ids)
      assert_equal Encoding::ASCII_8BIT, raw.encoding
      assert_equal "ab".b, raw
    ensure
      pipe.close
    end
  end

  def test_block_form_closes_pipeline
    pipe_ref = nil
    Ztok::Pipeline.byte_id do |pipe|
      pipe_ref = pipe
      assert_equal [ord_x], pipe.encode("x")
    end
    refute_nil pipe_ref
    assert pipe_ref.closed?, "block-form should auto-close"
  end

  def ord_x
    "x".ord
  end
end
